samples/99_zil_interpreter/app/parsers/scanning_parser.rb

```diff
@@ -0,0 +1,168 @@
+class ScanningParser
+
+  def initialize(work_dir)
+    @work_dir = work_dir
+  end
+
+  def parse(args, file_name)
+    args.state.ztokens ||= {}
+    file = args.gtk.read_file(@work_dir + file_name)
+
+    buffer = ""
+    lines = []
+    file.each_line.each { |line|
+      lines << line
+      line.each_char { |c| buffer << c }
+    }
+
+    log "Scanning '" + file_name + "' ..."
+    scanner = Scanner.new(file_name, buffer)
+
+    i = 0
+    #while scanner.scan_to('<')
+      scanner.scan_to('<')
+      block = scanner.read_block
+
+      # block should be < to > statement, but some multiline tokens starting with ( to ) break the logic
+
+      #if block.include?('PALANTIR') == true # grep out PALANTIR objects to show to kfischer
+        log "----- " + file_name + " STATEMENT"+ i.to_s + " -----"
+        log block
+      #end
+
+      i += 1
+    #end
+
+    log "Scan completed."
+
+    #i = 0
+    #num_lines = lines.length
+    #while i < num_lines
+    #  i += 1
+    #end
+  end
+end
+
+class Scanner
+  def initialize(file_name, buffer)
+    @buffer = buffer
+    @file_name = file_name
+    reset
+  end
+
+  def reset
+    @buf_index = -1
+  end
+
+  def scan_to(scan_char)
+    found = false
+
+    while @buf_index < @buffer.length - 1
+      @buf_index += 1
+      buf_char = @buffer[@buf_index]
+
+      if buf_char == scan_char
+        #log "Found '" + scan_char + "' at " + @line_num.to_s + ":" + @line_index.to_s
+        found = true
+        break
+      end
+
+    end
+
+    found
+  end
+
+  def peek_next
+    @buffer[@buf_index + 1]
+  end
+
+  def peek_atom
+    atom_chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-+*/?="
+    buf_index = @buf_index
+    atom = ""
+    while buf_index < @buffer.length - 1 # iterate characters in line
+      buf_char = @buffer[buf_index]
+      buf_index += 1
+
+      if atom_chars.include?(buf_char)
+        #log "[TMP] " + buf_char
+        atom += buf_char
+      else
+        break
+      end
+    end
+    atom
+  end
+
+  def read_atom

```

Comment:
don't understand the need for both `peek_atom` and `read_atom` hope you'd be able to provide some clarification, ZIL IDs can only be comprised of `A-Z` and `-` if I read correctly,
seems a _bit_ bug prone

Reply:
Agreed.  Fixed in update.